src/cargo/sources/git/utils.rs

```diff
@@ -645,9 +671,28 @@ pub fn fetch(repo: &mut git2::Repository,
 
     debug!("doing a fetch for {}", url);
     with_fetch_options(&repo.config()?, url, config, &mut |mut opts| {
-        debug!("initiating fetch of {} from {}", refspec, url);
-        let mut remote = repo.remote_anonymous(url.as_str())?;
-        remote.fetch(&[refspec], Some(&mut opts), None)?;
+        let mut is_clean_try = false;
+        loop {
+            debug!("initiating fetch of {} from {}", refspec, url);
+            let res = repo.remote_anonymous(url.as_str())?
+                .fetch(&[refspec], Some(&mut opts), None);
+            let err = match res {
+                Ok(()) => break,
+                Err(e) => e,
+            };
+            debug!("fetch failed: {}", err);
+            match err.class() {
+                _ if is_clean_try => return Err(err.into()),
+                git2::ErrorClass::Reference => {}
+                _ => return Err(err.into()),
+            }
+            debug!("looks like this is a corrupt repository, reinitializing \
+                    and trying again");
+            if reinitialize(repo).is_err() {
+                return Err(err.into())
+            }
+            is_clean_try = true;
+        }
```

Comment:
It's not obvious why we'll exit this loop eventually...

Maybe let's do like a ten tries and then bail with an internal error?